README.md
Buildtime Trend as a Service
============================

Visualise what's trending in your build process

[![Buildtime Trend as a Service](http://img.shields.io/badge/release-v0.3-blue.svg)](https://github.com/buildtimetrend/service/releases/latest)
[![Buildtime Trend as a Service (dev)](http://img.shields.io/badge/dev-v0.4.dev-blue.svg)](https://github.com/buildtimetrend/service/zipball/master)

[![Build Status](https://travis-ci.org/buildtimetrend/service.svg?branch=master)](https://travis-ci.org/buildtimetrend/service)
[![Code Health](https://landscape.io/github/buildtimetrend/service/master/landscape.svg)](https://landscape.io/github/buildtimetrend/service/master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/buildtimetrend/service/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/buildtimetrend/service/?branch=master)
[![Codacy Badge](https://www.codacy.com/project/badge/4ff71ed6b542493ab6f407f4c37aeb01)](https://www.codacy.com/public/ruleant/service)
[![Code Issues](https://www.quantifiedcode.com/api/v1/project/88a900bee57a41ec82a3fd75e8e2bb0a/badge.svg)](https://www.quantifiedcode.com/app/project/88a900bee57a41ec82a3fd75e8e2bb0a)
[![Stack Share](http://img.shields.io/badge/tech-stack-0690fa.svg)](http://stackshare.io/ruleant/buildtime-trend)

[![Buildtime trend](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/latest)](https://buildtimetrend-dev.herokuapp.com/dashboard/buildtimetrend/service/index.html)
[![Total builds](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/builds/month)](https://buildtimetrend-dev.herokuapp.com/dashboard/buildtimetrend/service/index.html)
[![Percentage passed build jobs](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/passed/month)](https://buildtimetrend-dev.herokuapp.com/dashboard/buildtimetrend/service/index.html)
[![Days since last failed build job](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/last_fail)](https://buildtimetrend-dev.herokuapp.com/dashboard/buildtimetrend/service/index.html)

This project contains the files that offer Buildtime Trend as a service. The [Buildtime Trend library](https://github.com/buildtimetrend/python-lib) powers this service.

## Usage

### Use the service available on Heroku

- production : https://buildtimetrend.herokuapp.com/

    Currently in beta.

    Available for free for Open Source projects (public on GitHub), thanks to [Keen.io](https://keen.io) for hosting the timing data.

    Private projects will be supported soon.

- development : https://buildtimetrend-dev.herokuapp.com/

    Used for development, expect frequent changes, accepts only a limited number of projects.
    [Contact us](#Contact) if you'd like to test the development version. It's recommended to use the production version.

### Clone the project and deploy it to Heroku

See [instructions on the wiki](https://github.com/buildtimetrend/python-lib/wiki/Running-service-on-Heroku) on how to configure the service on Heroku.

### Run it on your own server

- Launch a CherryPy instance hosting the service :

```bash
python service.py
```

> **Remark :** By default it will attach to port 5000, use environment variable `PORT` to override this.

More [configuration options](#config-file) are available.

- To enable a worker to handle the task queue (optional) :

```bash
utils/start_worker.sh
```

A Celery compatible task queue service is required to use the task queue, see the `task_queue` config parameter in the [configuration options](#config-file) section.

## Components

- [Index page](#index-page)
- [Dashboard](#dashboard)
- [Service usage statistics](#service-usage-statistics)
- [Shield badges](#shield-badges)
- [Process Travis CI build log](#process-travis-ci-build-log)

### Index page
- path : `/`

### Dashboard

Display a dashboard with Buildtime Trend charts

- path : `/dashboard`
- usage : `/dashboard/repo_owner/repo_name?refresh=<refresh_rate>&timeframe=<timeframe_setting>`
- parameters :
  - `repo_owner` : name of the Github repo owner, fe. `buildtimetrend`
  - `repo_name` : name of the Github repo, fe. `service`
  - `refresh` (optional) : enables auto-refreshing the dashboard charts. 0 = disabled (default), a positive integer is the number of minutes after which the charts are refreshed, a value of 10 will refresh the charts every 10 minutes. The refresh rate should at least be equal to maximum age of the Query cache, if it is less, the cache max age value will be used (typically, 10 min).
  - `timeframe` (optional) : the default timeframe can be set with this url parameter. Possible values : `day`, `week`, `month`, `year`. If it is not defined the default timeframe is used (`week`)
  - `filter_*` : set the default filter value. When this url parameter is defined, the corresponding filter dropdown will be set to this value and it will be applied to the queries of all charts and metrics. Available filters :
    - `filter_build_matrix` : Build matrix environment settings, usually a combination of language, language version and operating system : fe. `python 2.7 linux`
    - `filter_build_result` : Build job result, possible values : `passed`, `errored`, `failed`, ...
    - `filter_build_trigger` : What triggered the build job, possible values : `pull_request`, `push` (git push)
    - `filter_branch` : Repository branch that was build : fe. `master`


> **Remark :** When visiting `/dashboard` (without parameters), an overview of all hosted projects is displayed.

### Service usage statistics

Display a dashboard with Buildtime Trend service usage statistics

- path : `/stats`
- usage : `/stats?timeframe=<timeframe_setting>`
- parameters :
  - `timeframe` (optional) : the default timeframe can be set with this url parameter. Possible values : `day`, `week`, `month`, `year`. If it is not defined the default timeframe is used (`week`)

### Shield badges

Generate a shield badges

- path : `/badge`
- usage : `/badge/repo_owner/repo_name/badge_type/interval`
- parameters :
  - `repo_owner` : name of the Github repo owner, fe. `buildtimetrend`
  - `repo_name` : name of the Github repo, fe. `service`
  - `badge_type` : type of badge, options :
    - `latest` : buildtime of last build job
    - `avg` : average buildtime of buildjobs in period set by `interval` (default)
    - `jobs` : number of build jobs in period set by `interval`
    - `builds` : number of builds in period set by `interval`
    - `passed` : percentage of successful build jobs during `interval`
    - `last_fail` : days since last failed build job
  - `interval` : time interval, options :
    - `week` (default) : events of last week (last 7 days)
    - `month` : events of last month (last 30 days)
    - `year` : events of last year (last 52 weeks)

#### Badge examples
- ![Latest build duration](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/latest) : `/badge/buildtimetrend/service/latest`
- ![Average buildtime (week)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/avg/week) : `/badge/buildtimetrend/service/avg/week`
- ![Average buildtime (month)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/avg/month) : `/badge/buildtimetrend/service/avg/month`
- ![Average buildtime (year)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/avg/year) : `/buildtimetrend/service/avg/year`
- ![Build jobs (week)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/jobs/week) : `/badge/buildtimetrend/service/jobs/week`
- ![Build jobs (month)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/jobs/month) : `/badge/buildtimetrend/service/jobs/month`
- ![Build jobs (year)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/jobs/year) : `/buildtimetrend/service/jobs/year`
- ![Builds (week)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/builds/week) : `/badge/buildtimetrend/service/builds/week`
- ![Builds (month)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/builds/month) : `/badge/buildtimetrend/service/builds/month`
- ![Builds (year)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/builds/year) : `/buildtimetrend/service/builds/year`
- ![% passed build jobs (week)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/passed/week) : `/badge/buildtimetrend/service/passed/week`
- ![% passed build jobs (month)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/passed/month) : `/badge/buildtimetrend/service/passed/month`
- ![% passed build jobs (year)](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/passed/year) : `/badge/buildtimetrend/service/passed/year`
- ![days since last failed build job](https://buildtimetrend-dev.herokuapp.com/badge/buildtimetrend/service/last_fail) : `/badge/buildtimetrend/service/last_fail`

### Process Travis CI build log

Loads a Travis CI build log file, processes it and sends the data to Keen.io.

- path : `/travis`
- usage : `/travis/<repo_owner>/<repo_name>/<first_build>/<last_build>`
- parameters :
  - `repo_owner` : name of the Github repo owner, fe. `buildtimetrend`
  - `repo_name` : name of the Github repo, fe. `service`
  - `first_build` : Travis CI build number of the first build to be processed
  - `last_build` : (optional) Travis CI build number of the last build to be processed

If `last_build` is not defined, only a single build will be processed, namely the one defined by `first_build`.
If `last_build` is defined, it turns into batch mode : multiple builds will be processed : all builds from `first_build` to and including `last_build` will be scheduled to be processed.
To limit the load on the worker, a delay is added to every next build. F.e. build #1 will be executed immediately, build #2 after x seconds, build #3 after x\*2 seconds, build #n after x\*(n-1) seconds.
By default the delay is 3 seconds, but this can be configured with the `multi_import`.`delay` setting in `config_service.yml` or `BTT_MULTI_DELAY` env variable.
The maximum number of builds that can be imported at once in batch mode is by default limited to 100. This can be configured with the `multi_import`.`max_builds` setting in `config_service.yml` or `BTT_MULTI_MAX_BUILDS` env variable.

OR

  - `payload` : Travis CI notification payload, more info in the [Travis CI documentation](http://docs.travis-ci.com/user/notifications/#Webhook-notification)

To trigger the service at the end of a Travis CI build, add this to your `.travis.yml` file:

```yaml
    notifications:
      webhooks:
        # trigger Buildtime Trend Service to parse Travis CI log and send result to Keen.io
        - https://buildtimetrend.herokuapp.com/travis
```


ChangeLog
v0.4 (not yet released)
- issue #38 : add 'days since last failed build job' badge
- issue #146 : use url of badges of local service instance
- issue #145 : upgrade to CherryPy 5.0.1

v0.3 (released on 18Nov2015)
- issue #82 : use generator to return updates of TravisParser
- issue #83 : check if build exists in database before submitting
- add `denied_repo` setting
- TravisData.process_build_jobs is a generator
- issue #21 : [dashboard] add charts with build job results
- issue #52 : [dashboard] upgrade to Keen JS API 3.2
- issue #39 : format buildtime duration badge
- issue #23 : [dashboard] add charts with build job results grouped by branch or by build
environment parameters
- issue #92 : fix : duration formatting fails when time is undefined
- issue #70 : [dashboard] use bootstrap buttons and ButtonClass on dashboard to select timeframe
- move is_repo_allowed(), check_process_parameters() and format_duration() from service to library
- issue #64 : Use queue and worker in heroku
- issue #84 : Add Service statistics page
- issue #99 : Enable auto refreshing dashboard charts with `refresh` url parameter
- issue #89 : add batch mode to process multiple builds
- issue #95 : don't use pickle as the celery serialiser, due to security concerns
- issue #113 : Set default timeframe with `timeframe` url parameter
- issue #121 : use url parameter to set query filters

v0.2 (released on 18Feb2015)
- issue #4, #51 : Use heroku to process Travis logs asynchronously
- set loglevel with environment variable BTT_LOGLEVEL and config file setting
- issue #17 : add build status badges
- issue #74 : return custom favicon.ico and robots.txt
- issue #71 : host dashboards of projects on service
- schema change : save build job data in build_jobs.job (was builds.build)
- issue #77 : add apple-touch-icon.png

This project follows the version numbering of the other modules of the
Buildtime Trend project. The service was introduced in version v0.2, hence the
first release starts with v0.2


service.py
#!/usr/bin/env python
# vim: set expandtab sw=4 ts=4:
"""
Buildtime Trend as a Service Cherrypy handlers.

Service components :
- /dashboard : display Buildtime Trend dashboard of hosted projects
- /stats     : display Buildtime Trend dashboard with service usage statistics
- /badge     : generate badges with metrics of project build data
- /travis    : retrieve and parse build data from Travis CI

Copyright (C) 2014-2016 Dieter Adriaenssens <ruleant@users.sourceforge.net>

This file is part of buildtimetrend/service
<https://github.com/buildtimetrend/service/>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
"""

import os
import cgi
import cherrypy
from cherrypy.process.plugins import PIDFile
import urllib
import constants
from buildtimetrend.travis.tools import process_notification_payload
from buildtimetrend.travis.tools import check_authorization
from buildtimetrend.settings import Settings
from buildtimetrend import logger
from buildtimetrend.tools import get_repo_slug
from buildtimetrend.tools import check_file
from buildtimetrend.tools import file_is_newer
from buildtimetrend import keenio
from buildtimetrend.keenio import check_time_interval
from buildtimetrend.keenio import get_avg_buildtime
from buildtimetrend.keenio import get_total_builds
from buildtimetrend.keenio import get_pct_passed_build_jobs
from buildtimetrend.keenio import get_result_color
from buildtimetrend.keenio import get_total_build_jobs
from buildtimetrend.keenio import get_latest_buildtime
from buildtimetrend.keenio import get_all_projects
from buildtimetrend.dashboard import get_config_string
from buildtimetrend.service import is_repo_allowed
from buildtimetrend.service import format_duration
from buildtimetrend.service import validate_travis_request
from celery_worker import is_worker_enabled
import tasks

SERVICE_WEBSITE_LINK = "<a href='https://buildtimetrend.github.io/service'>" \
                       "Buildtime Trend as a Service</a>"
TRAVIS_URL = '/travis'
ASSETS_URL = '/assets'
DASHBOARD_URL = '/dashboard'
STATS_URL = '/stats'
BADGE_URL = '/badge'
ROOT_DIR = os.path.dirname(os.path.realpath(__file__))
STATIC_DIR = os.path.join(ROOT_DIR, 'static')
DASHBOARD_DIR = os.path.join(STATIC_DIR, 'dashboard')
ASSETS_DIR = os.path.join(DASHBOARD_DIR, 'assets')
IMAGES_DIR = os.path.join(ASSETS_DIR, 'images')
FAVICON_PATH = os.path.join(IMAGES_DIR, 'favicon.ico')
APPLE_ICON_PATH = os.path.join(IMAGES_DIR, 'apple-touch-icon.png')
APPLE_ICON_PRECOMPOSED_PATH = os.path.join(IMAGES_DIR,
                                           'apple-touch-icon-precomposed.png')
ROBOTS_PATH = os.path.join(STATIC_DIR, 'robots.txt')


class Dashboard(object):

    """Buildtime Trend Dashboard handler."""

    def __init__(self):
        """Constructor."""
        self.settings = Settings()
        self.logger = logger

        self.file_projects = os.path.join(DASHBOARD_DIR, "projects.html")
        self.file_projects_service = os.path.join(
            DASHBOARD_DIR, "projects_service.html"
        )
        self.file_index = os.path.join(DASHBOARD_DIR, "index.html")
        self.file_index_service = os.path.join(
            DASHBOARD_DIR, "index_service.html"
        )

    @cherrypy.expose
    def index(self):
        """Index page with overview of hosted projects."""
        # Create project overview for Buildtime Trend as a Service :
        # if it doesn't exist, or if it is older than the file from
        # which it is generated
        if modify_index(self.file_projects, self.file_projects_service):
            return open(self.file_projects_service)
        else:
            raise cherrypy.HTTPError(404, "File not found")

    def dashboard(self):
        """Dashboard page."""
        # Create dashboard index for Buildtime Trend as a Service :
        # if it doesn't exist, or if it is older than the file from
        # which it is generated
        if modify_index(self.file_index, self.file_index_service):
            return open(self.file_index_service)
        else:
            raise cherrypy.HTTPError(404, "File not found")

    @cherrypy.expose
    def default(self, repo_owner=None, repo_name=None, page="",
                refresh=None, timeframe=None,
                filter_build_matrix=None, filter_build_result=None,
                filter_build_trigger=None, filter_branch=None):
        """
        Default page.

        Returns index page or config file,
        redirects index page in all other cases.

        Parameters :
        - repo_owner : name of the Github repo owner, fe. `buildtimetrend`
        - repo_name : name of the Github repo, fe. `service`
        - page : requested page (index.html or config.js)
        """
        if page == "index.html" or repo_owner is None and repo_name is None:
            # display index page
            return self.dashboard()
        elif page == "config.js":
            # display config file
            return self.config_js(repo_owner, repo_name)
        else:
            # redirect to index page, if no other page matches
            if repo_owner is None or repo_name is None:
                url = DASHBOARD_URL
            else:
                repo_slug = get_repo_slug(repo_owner, repo_name)
                url = "{0!s}/{1!s}/index.html".format(
                    DASHBOARD_URL, cgi.escape(repo_slug)
                )

                # add url parameters
                url_params = {}
                if refresh is not None:
                    url_params['refresh'] = refresh
                if timeframe is not None:
                    url_params['timeframe'] = timeframe
                if filter_build_matrix is not None:
                    url_params['filter_build_matrix'] = filter_build_matrix
                if filter_build_result is not None:
                    url_params['filter_build_result'] = filter_build_result
                if filter_build_trigger is not None:
                    url_params['filter_build_trigger'] = filter_build_trigger
                if filter_branch is not None:
                    url_params['filter_branch'] = filter_branch

                if url_params:
                    url = "{0!s}?{1!s}".format(
                        url, urllib.urlencode(url_params)
                    )

            # rewrite url
            raise cherrypy.HTTPRedirect(url)

    @cherrypy.expose
    def config_js(self, repo_owner=None, repo_name=None):
        """
        Config file for dashboard.

        Parameters :
        - repo_owner : name of the Github repo owner, fe. `buildtimetrend`
        - repo_name : name of the Github repo, fe. `service`
        """
        # define extra settings
        extra = {
            'serviceUrl': "/"  # use this service instance for badge generation
        }

        repo = get_repo_slug(repo_owner, repo_name)

        # Check if repo is allowed
        if repo is not None and not is_repo_allowed(repo):
            message = "Project '{}' is not allowed."
            self.logger.info(message.format(repo))
            extra['message'] = message.format(cgi.escape(repo))
            repo = None
        else:
            self.logger.info(
                "Generated dashboard config for project '%s'", repo
            )

        # add project list
        extra.update(get_config_project_list())

        # return config file
        return get_config_string(repo, extra)


class Stats(object):

    """Service usage stats page handler."""

    def __init__(self):
        """Constructor."""
        self.settings = Settings()
        self.logger = logger

        self.file_stats = os.path.join(DASHBOARD_DIR, "stats.html")
        self.file_stats_service = os.path.join(
            DASHBOARD_DIR, "stats_service.html"
        )

    @cherrypy.expose
    def index(self, timeframe=None, count=None):
        """Service usage stats page."""
        # Create usage stats page for Buildtime Trend as a Service,
        # if it doesn't exist, or if it is older than the file from
        # which it is generated
        if modify_index(self.file_stats, self.file_stats_service):
            return open(self.file_stats_service)
        else:
            raise cherrypy.HTTPError(404, "File not found")

    @cherrypy.expose
    def config_js(self):
        """Config file for usage stats dashboard."""
        # define extra settings
        extra = {
            'projectName': "Service usage stats"
        }

        # add project list
        extra.update(get_config_project_list())

        # return config file
        return get_config_string(None, extra)


class Badges(object):

    """Generate shield badges."""

    def __init__(self):
        """Initialise class."""
        self.settings = Settings()

        # get logger
        self.logger = logger

    @cherrypy.expose
    def default(self, repo_owner=None, repo_name=None, badge_type="avg",
                interval=None):
        """
        Generate a shield badge.

        Parameters :
        - repo_owner : name of the Github repo owner, fe. `buildtimetrend`
        - repo_name : name of the Github repo, fe. `service`
        - badge_type : type of badge, options :
          - latest : buildtime of last build job
          - avg : average buildtime of buildjobs in period set by `interval`
                  (default)
          - jobs : number of build jobs in period set by `interval`
          - builds : number of builds in period set by `interval`
          - passed : percentage of successful build jobs during `interval`
        - interval : time interval, options :
          - week (default) : events of last week (last 7 days)
          - month : events of last month (last 30 days)
          - year : events of last year (last 52 weeks)
        """
        # parameter check
        repo = get_repo_slug(repo_owner, repo_name)
        badge_type = str(badge_type).lower()
        interval = check_time_interval(interval)["name"]

        badge_subject = "buildtime"
        badge_status = "trend"
        badge_colour = "blue"
        format_string = "{:d}"

        if repo is not None and is_repo_allowed(repo):
            if badge_type == "latest":
                # get last duration
                value = get_latest_buildtime(repo)
                badge_status = format_duration(value)
            elif badge_type == "jobs":
                badge_subject = "{0!s}_({1!s})".format(badge_type, interval)
                value = get_total_build_jobs(repo, interval)
                format_string = "{:d}"
                badge_status = format_string.format(value)
            elif badge_type == "builds":
                badge_subject = "{0!s}_({1!s})".format(badge_type, interval)
                value = get_total_builds(repo, interval)
                format_string = "{:d}"
                badge_status = format_string.format(value)
            elif badge_type == "passed":
                badge_subject = "{0!s}_({1!s})".format(badge_type, interval)
                value = get_pct_passed_build_jobs(repo, interval)
                badge_colour = get_result_color(value, 100, 75)
                format_string = "{:d}%"
                badge_status = format_string.format(value)
            elif badge_type == "last_fail":
                badge_subject = "{0!s}".format((badge_type))
                value = keenio.get_days_since_fail(repo)
                badge_colour = get_result_color(value, 3, 0)
                format_string = "{:d} days"
                badge_status = format_string.format(value)
            else:
                # calculate average
                badge_subject = "{0!s}_({1!s})".format(badge_subject, interval)
                value = get_avg_buildtime(repo, interval)
                badge_status = format_duration(value)

            # valid duration is 0 or greater int or float
            if type(value) not in (float, int) or value < 0:
                badge_status = "unknown"
                badge_colour = "lightgrey"

            self.logger.info(
                "Badge type %s (interval : %s) for %s, value : %s",
                badge_type, interval, repo, badge_status)

        # Redirect to shields.io API to generate badge
        raise cherrypy.HTTPRedirect(
            "https://img.shields.io/badge/{0!s}-{1!s}-{2!s}.svg".format(
                badge_subject, badge_status, badge_colour
            )
        )


class Root(object):

    """Root handler."""

    def __init__(self):
        """
        Initialise class.

        Load config file and set loglevel, define error page handler
        """
        self.settings = Settings()
        self.settings.load_settings(config_file=constants.CONFIG_FILE)
        self.settings.set_client(
            constants.CLIENT_NAME,
            constants.CLIENT_VERSION
        )

        self.file_index = os.path.join(STATIC_DIR, "index.html")

        cherrypy.config.update({'error_page.404': self.error_page_404})

        # get logger
        self.logger = logger

    @cherrypy.expose
    def index(self):
        """Index page."""
        if check_file(self.file_index):
            return open(self.file_index)
        else:
            raise cherrypy.HTTPError(404, "File not found")

    def error_page_404(self, status, message, traceback, version):
        """Error Page (404)."""
        self.logger.error(
            "Cherrypy %s : Error loading page (%s) : %s\nTraceback : %s",
            version, status, message, traceback
        )
        return "This page doesn't exist, please check usage on " \
               "the {} website.".format(SERVICE_WEBSITE_LINK)


class TravisParser(object):

    """
    Travis CI build timing and build data parser.

    Retrieve timing data from Travis CI, parse it and store it in Keen.io.
    """

    def __init__(self):
        """Initialise class."""
        self.settings = Settings()

        # get logger
        self.logger = logger

    @cherrypy.expose
    def default(self, repo_owner=None, repo_name=None,
                first_build=None, last_build=None,
                payload=None):
        """
        Default handler.

        Visiting this page triggers loading and processing the build log
        and data of a travis CI build process.

        If last_build is defined, all builds from first_build until last_build
        will be retrieved and processed.

        If only payload is defined, repo and build data
        will be extracted from the payload.

        Parameters:
        - repo_owner : name of the Github repo owner, fe. `buildtimetrend`
        - repo_name : name of the Github repo, fe. `service`
        - first_build : first build number to process (int)
        - last_build : last build number to process (int)
        - payload : Travis CI notification payload (json)
        """
        cherrypy.response.headers['Content-Type'] = 'text/plain'
        # reset settings
        self.settings.set_project_name(None)
        self.settings.add_setting('build', None)

        self.logger.debug(
            "Check Travis headers : %r",
            cherrypy.request.headers
        )

        repo = get_repo_slug(repo_owner, repo_name)

        # load parameters from the Travis notification payload
        if self.check_travis_notification():
            payload_params = process_notification_payload(payload)

            # assign payload parameters
            if repo is None and "repo" in payload_params:
                repo = payload_params["repo"]
            if first_build is None and last_build is None and \
                    "build" in payload_params:
                first_build = payload_params["build"]

        # check parameter validity, check returns error message
        # or None if parameters are valid
        params_valid = validate_travis_request(repo, first_build)
        if params_valid is not None:
            self.logger.warning(params_valid)
            return params_valid

        if last_build is None:
            self.logger.warning(
                "Request to process build #%s of repo %s", first_build, repo
            )

            # schedule task with 10 second delay to give Travis CI time
            # to add the finished_at property. (issue #96)
            return self.schedule_task(repo, first_build, 10)

        return self.multi_build(repo, first_build, last_build)

    def multi_build(self, repo, first_build, last_build):
        """
        Schedule processing multiple consecutive builds.

        All builds from first_build until last_build
        will be retrieved and processed.

        The total number of builds to be scheduled is limited by the
        `multi_import.max_builds` config parameter.

        Every next scheduled build will be delayed by the
        `multi_import.delay` config parameter.

        Parameters:
        - repo : repo name (fe. buildtimetrend/service)
        - first_build : first build number to process (int)
        - last_build : last build number to process (int)
        """
        first_build = int(first_build)
        last_build = int(last_build)
        message = ""
        multi_import = Settings().get_setting("multi_import")
        max_multi_builds = int(multi_import["max_builds"])

        if last_build < first_build:
            tmp_msg = "Warning : last_build should be equal" \
                " or larger than first_build"
            self.logger.warning(tmp_msg)
            message += tmp_msg + "\n"
            last_build = first_build

        if (last_build - first_build) > max_multi_builds:
            tmp_msg = "Warning : number of multiple builds is limited to {:d}"
            self.logger.warning(tmp_msg.format(max_multi_builds))
            message += tmp_msg.format(max_multi_builds) + "\n"
            last_build = first_build + max_multi_builds

        message += "Request to process build(s) #{first_build:d} to" \
            " #{last_build:d} of repo {repo}:\n".format(**locals())

        build = first_build
        delay = 0

        while build <= last_build:
            message += self.schedule_task(repo, build, delay) + "\n"
            delay += multi_import["delay"]
            build += 1

        return message

    def schedule_task(self, repo, build, delay=0):
        """
        Schedule task.

        Parameters:
        - repo : repo name (fe. buildtimetrend/service)
        - build : build number to process (int)
        - delay : delay before task should be started, in seconds
        """
        # process travis build
        if is_worker_enabled():
            task = tasks.process_travis_buildlog.apply_async(
                (repo, build), countdown=int(delay)
            )
            temp_msg = "Task scheduled to process build #{build}" \
                " of repo {repo} : {task_id}"
            self.logger.warning(
                temp_msg.format(build=build, repo=repo, task_id=task.id)
            )
            return temp_msg.format(
                build=cgi.escape(str(build)),
                repo=cgi.escape(str(repo)),
                task_id=cgi.escape(str(task.id))
            )
        else:
            return tasks.process_travis_buildlog(repo, build)

    def check_travis_notification(self):
        """
        Check Travis CI notification request.

        Load Authorization and Travis-Repo-Slug headers and check if
        the Authorization header is correct
        """
        if "Authorization" not in cherrypy.request.headers:
            self.logger.debug("Authorization header is not set")
            return False

        if "Travis-Repo-Slug" not in cherrypy.request.headers:
            self.logger.debug("Travis-Repo-Slug header is not set")
            return False

        return check_authorization(
            cherrypy.request.headers["Travis-Repo-Slug"],
            cherrypy.request.headers["Authorization"]
        )


def get_config_project_list():
    """
    Return a list of repoNames of projects hosted on this website.

    Create list of only allowed project repos
    and convert values from unicode to UTF8
    """
    allowed_projects = [
        str(x) for x in get_all_projects() if is_repo_allowed(x)
    ]

    # check if collection is empty
    if allowed_projects:
        return {'projectList': allowed_projects}
    else:
        return {}


def modify_index(file_original, file_modified):
    """
    Modify html file for Buildtime Trend as a Service.

    Adjust paths to 'assets' :
    the relative path is changed to an absolute path.

    Parameters:
    - file_original : Path of the original file
    - file_modified : Path of the modified file hosted on the service
    """
    if not file_is_newer(file_modified, file_original):
        with open(file_original, 'r') as infile, \
                open(file_modified, 'w') as outfile:
            for line in infile:
                line = line.replace("assets", ASSETS_URL)
                outfile.write(line)

    if check_file(file_modified):
        logger.info(
            "Created index service file : %s",
            file_modified
        )
        return True
    else:
        return False


if __name__ == "__main__":
    # write pid to a file
    PIDFile(cherrypy.engine, '/tmp/btt_service.pid').subscribe()

    # configure cherrypy webserver host and port
    # inspired by https://github.com/craigkerstiens/cherrypy-helloworld
    cherrypy.config.update({
        'server.socket_host': '0.0.0.0',
        'server.socket_port': int(os.environ.get('PORT', '5000')),
        'engine.autoreload.on': False
    })

    ROOT_CONFIG = {
        '/favicon.ico': {
            'tools.staticfile.on': True,
            'tools.staticfile.filename': FAVICON_PATH
        },
        '/apple-touch-icon.png': {
            'tools.staticfile.on': True,
            'tools.staticfile.filename': APPLE_ICON_PATH
        },
        '/apple-touch-icon-precomposed.png': {
            'tools.staticfile.on': True,
            'tools.staticfile.filename': APPLE_ICON_PRECOMPOSED_PATH
        },
        '/robots.txt': {
            'tools.staticfile.on': True,
            'tools.staticfile.filename': ROBOTS_PATH
        },
        ASSETS_URL: {
            'tools.staticdir.on': True,
            'tools.staticdir.dir': ASSETS_DIR
        }
    }

    # assign handlers to entry paths
    cherrypy.tree.mount(Root(), '/', ROOT_CONFIG)
    cherrypy.tree.mount(Dashboard(), DASHBOARD_URL)
    cherrypy.tree.mount(Stats(), STATS_URL)
    cherrypy.tree.mount(Badges(), BADGE_URL)
    cherrypy.tree.mount(TravisParser(), TRAVIS_URL)

    # start server
    cherrypy.engine.start()
    cherrypy.engine.block()
